File: Algorithm/back_tracing/Combination Sum/Combination Sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> templist;
        vector<vector<int>>list;
        int start=0;
        int remain=target;
        backtrace(list,templist,candidates,remain,start);
        return list;
        
    }
    void backtrace(vector<vector<int>> &list,vector<int>templist,vector<int>candidatates,int remain,int start)
    {
        if(remain<0) return;
        if(remain==0) list.push_back(templist);
        for(int i=start;i<candidatates.size();i++)
        {   
            
            templist.push_back(candidatates[i]);
            backtrace(list,templist,candidatates,remain-candidatates[i],i);
            templist.pop_back();
        }

    }    
};

File: Algorithm/back_tracing/Combination Sum/combinationSum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()
        lst = []
        n = len(candidates)
        def backtrace(i,tep,remain):
            if remain < 0: return
            elif  remain == 0:
                lst.append(tep)
                return 
            for j in range(i,n):
                if remain - candidates[j]<0:break
                backtrace(j,tep+[candidates[j]],remain-candidates[j])
        backtrace(0,[],target)
        return lst
